### PR TITLE
envydis/falcon: updates and documentation for v5 and v6 opcodes

### PR DESCRIPTION
**Pull request by hexkyz**
This PR provides:
- Support for all co-processor opcodes;
- Names for all previously unknown Falcon v5 opcodes;
- Names for the new Falcon v6 SPRs and 2 new unknown opcodes.

This intends to synchronize envydis' source with the research ongoing at https://switchbrew.org/wiki/TSEC.

### CODE REVIEW

**Comment by mwkmwkmwk** [comment]
Please add indirect versions of all new crypt opcodes — they can just be interpolated in the obvious way

Other than that, LGTM

**Comment by hexkyz** [comment]
Added indirect versions for all the crypt opcodes as requested by @mwkmwkmwk.

**Review by mwkmwkmwk — APPROVED** [review]

File: envydis/falcon.c [hunk 1]
```diff
@@ -256,18 +259,21 @@ static struct insn tabcocmd[] = {
...
 	{ 0x98000000, 0xfc000000, N("cs0exec"), CIMM2 },
 	/* like above, but operates on another script slot */
 	{ 0x9c000000, 0xfc000000, N("cs1begin"), CIMM2 },
 	{ 0xa0000000, 0xfc000000, N("cs1exec"), CIMM2 },
+	{ 0xa8000000, 0xfc000000, N("cchmod"), CREG1, CIMM2 },
 	{ 0xac000000, 0xfc000000, N("cxor"), CREG1, CREG2 },
 	{ 0xb0000000, 0xfc000000, N("cadd"), CREG1, CIMM2 }, /* add immediate to register, modulo 2^64 */
 	{ 0xb4000000, 0xfc000000, N("cand"), CREG1, CREG2 },
 	{ 0xb8000000, 0xfc000000, N("crev"), CREG1, CREG2 }, /* reverse bytes */
-	{ 0xbc000000, 0xfc000000, N("cprecmac"), CREG1, CREG2 }, /* shift left by 1. if bit shifted out the left side was set, xor with 0x87. */
```
Comment (saper) [inline envydis/falcon.c:270]: Renaming this breaks assembly of this [arguably ancient code](https://github.com/polachok/pscnv/blob/05fc1a32334e560fe52bfc2d73290952465f4b50/pscnv/nv98_crypt.fuc#L621)

    pscnv/pscnv/nv98_crypt.fuc:621.3-622.1: No match

Changing `cprecmac` to `cgfmul` in the nv98_crypt.fuc file fixes it, it took me a while to figure out that "No match" means "Illegal instruction" here.

Not this change, but also  [options after positional parameter CLI syntax](https://github.com/polachok/pscnv/blob/05fc1a32334e560fe52bfc2d73290952465f4b50/pscnv/CMakeLists.txt#L14-L15) got broken in the mean time.

Comment (mupuf) [reply envydis/falcon.c:270]: Oh boy, you are resurrecting ancient code.

Not sure the assembler ever guaranteed backwards compatibility with previous versions. It really never felt worth it AFAIK, especially for a reverse engineered ISA.

Mind telling us more about what you are doing?

Comment (saper) [reply envydis/falcon.c:270]: I am just learning things. I figured out that a smaller pscnv might be easier to study. Feel free to ignore this comment, I just thought somebody might be running into this problem and searching for an answer.

Now I wonder if the security co-processor can be made to do something useful unrelated to the GPU, but this is another story.

Comment (saper) [reply envydis/falcon.c:270]: Just found this: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/gpu/drm/nouveau/nvkm/engine/sec/fuc/g98.fuc0s?id=c0e297dc61f8d4453e07afbea1fa8d0e67cd4a34#n621 this is included with the newest linux kernel :)

**Pull request merged**